mail auth: Add Client-Port

u5surf opened the pull request:
### Proposed changes

Describe the use case and detail of the change.

#205 

When mail proxy with auth asks the HTTP server for authentication, it can add Client-Port in the request headers.

It can be observed `tcpdump` during test which is executed by prove [./mail_smtp.t ](https://github.com/nginx/nginx-tests/blob/master/mail_smtp.t) as below.

```
sudo tcpdump -i lo -A dst port 8080

GET /mail/auth HTTP/1.0
Host: 127.0.0.1
Auth-Method: none
Auth-User:
Auth-Pass:
Auth-Protocol: smtp
Auth-Login-Attempt: 1
Client-IP: 127.0.0.1
Client-Port: 46500
Client-Host: [UNAVAILABLE]
Auth-SMTP-Helo: example.com
Auth-SMTP-From: MAIL FROM:<test@example.com> SIZE=100
Auth-SMTP-To: RCPT TO:<example.com>
```

If this pull request addresses an issue on GitHub, make sure to reference that
issue using one of the
[supported keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).

Before creating a pull request, make sure to comply with the
[Contributing Guidelines](https://github.com/nginx/nginx/blob/master/CONTRIBUTING.md).

u5surf committed:
Mail: Add Client-Port in auth header.

in order to send it to authentication script.

ac000 commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1228,6 +1228,8 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
 #endif
 
     cscf = ngx_mail_get_module_srv_conf(s, ngx_mail_core_module);
+    ngx_uint_t client_port;
```
Please declare variables at top of scope (watch variable ordering and alignment).

ac000 commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1244,9 +1246,12 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
           + ahcf->header.len
           + sizeof(CRLF) - 1;
 
+
```
I don't think we need this extra blank line.

u5surf commented:
@ac000 Thanks reviewing. I fixed them.

ac000 commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1133,6 +1133,7 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
+    ngx_uint_t                 client_port;
     ngx_buf_t                 *b;
     ngx_str_t                  login, passwd;
     ngx_connection_t          *c;
```
Seeing as we use 'Christmas Tree' ordering for variable declarations, the `ngx_uint_t` variable should come between `ngx_connection_t` and `ngx_str_t`.

Thanks.

u5surf replied:
@ac000 Thanks, I fixed it

pluknet reviewed (changes requested):
What is the purpose of the proposed change? 
This info is missing in the commit log.

Note that unlike "Client-IP" which is used to make auth decision based on client address, port doesn't serve such purpose.

pluknet commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1133,6 +1133,7 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
     ngx_mail_auth_http_conf_t *ahcf)
 {
     size_t                     len;
+    ngx_uint_t                 client_port;
```
client_port is a single-use variable, does it make sense to introduce it?

ac000 replied:
If I was writing it I would probably just call `ngx_inet_get_port()` from `ngx_sprintf()`

u5surf replied:
> If I was writing it I would probably just call ngx_inet_get_port() from ngx_sprintf()

@ac000 Thanks, I fixed so

pluknet commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1244,6 +1246,8 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
                 + sizeof(CRLF) - 1
           + sizeof("Client-Host: ") - 1 + s->host.len + sizeof(CRLF) - 1
+                + sizeof(CRLF) - 1
+          + sizeof("Client-Port: ") - 1 + sizeof("65535") -1 + sizeof(CRLF) - 1
```
The change looks wrong, it adds an extra CRLF for seemingly no purpose.
Besides, it has whitespace issue.

ac000 replied:
Disregard

~~Everything else seems to add it...~~

u5surf replied:
@pluknet I got rid of extra CRLF. But I didn’t see what you said white space issue.

ac000 replied:
Maybe the ` -1 ` (after the second sizeof()) which should probably be ` - 1`

u5surf replied:
fixed

pluknet commented on src/mail/ngx_mail_auth_http_module.c:
```diff
@@ -1341,6 +1345,8 @@ ngx_mail_auth_http_create_request(ngx_mail_session_t *s, ngx_pool_t *pool,
                        s->connection->addr_text.len);
     *b->last++ = CR; *b->last++ = LF;
+    b->last = ngx_sprintf(b->last, "Client-Port: %ui" CRLF,
+                          client_port);
```
need to separate by an empty line, as this adds a new header

pluknet reviewed (commented):
Please avoid introducing a new prefix in the log summary.

See related commits for the hint on commit log style.

u5surf commented:
@pluknet 
Thanks reviewing.

> What is the purpose of the proposed change?
This info is missing in the commit log.
> 
> Note that unlike "Client-IP" which is used to make auth decision based on client address, port doesn't serve such purpose.

The reason I’ve added it so simple. I would like to resolve this issue #205. Of course I comprehend it is unnecessary to authenticate, but it seems to be just necessary to use for logging authentication script side. 
Maybe we satisfies something like a variable which can be specified in `auth_http_header` directive.

I fixed some your suggestions and commit message.

pluknet commented:
> @pluknet Thanks reviewing.
> 
> > What is the purpose of the proposed change?
> > This info is missing in the commit log.
> > Note that unlike "Client-IP" which is used to make auth decision based on client address, port doesn't serve such purpose.
> 
> The reason I’ve added it so simple. I would like to resolve this issue #205. Of course I comprehend it is unnecessary to authenticate, but it seems to be just necessary to use for logging authentication script side. Maybe we satisfies something like a variable which can be specified in `auth_http_header` directive.
> 
> I fixed some your suggestions and commit message.

Ok I see, thank you for clarification.

So the reason is not quite justified.
As such, I would abstain from adding the port as this will unnecessarily complicate the logic,
until further rationale appears behind this change.

Note that the logging is not required for making the auth decision, instead it is supposed to be made by the mail server itself.
